File: Difficulty: Medium/Longest Subarray with Sum K/longest-subarray-with-sum-k.py
class Solution:
    def longestSubarray(self, arr, k):  
        # code here
        prefix={}
        maxlen=0
        s=0
        for i in range(len(arr)):
            s+=arr[i]
            
            if s==k:
                maxlen=i+1
                
            if (s-k) in prefix:
                maxlen=max(maxlen,i-prefix[s-k])
            
            if s not in prefix:
                prefix[s]=i
                
        return maxlen

File: Difficulty: Medium/Longest Subarray with Sum K/longest-subarray-with-sum-k.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @returns {number}
 */
class Solution {
    longestSubarray(arr, k) {
        // code here
        let prefix={};
        let maxLen=0;
        let sum=0;
        for(let i=0;i<arr.length;i++){
            sum+=arr[i];
            if(sum==k){
                maxLen=i+1;
            }
            
            let required=sum-k;
            if(required in prefix){
                maxLen=Math.max(maxLen,i-prefix[required]);
            }
            
            if(!(sum in prefix)){
                prefix[sum]=i;
            }
        }
        return maxLen;
    }
}
